domain/application/service/service.go

```diff
@@ -519,6 +520,71 @@ func (s *WatchableService) WatchUnitAddressesHash(ctx context.Context, unitName
 	)
 }
 
+// WatchUnitInsertDeleteOnMachine returns a watcher that observes changes to the
+// units on a specified machine, emitting the names of the units. That is, we
+// emit unit names only when a unit is created or deleted on the specified machine.
+// The following errors may be returned:
+// - [applicationerrors.MachineNotFound] if the machine does not exist
+func (s *WatchableService) WatchUnitInsertDeleteOnMachine(ctx context.Context, machineName machine.Name) (watcher.StringsWatcher, error) {
+	desiredNetNodeUUID, err := s.st.GetMachineNetNodeUUIDFromName(ctx, machineName)
+	if err != nil {
+		return nil, errors.Errorf("getting net node uuid for machine %q: %w", machineName, err)
+	}
+
+	unitNamesOnMachineCache := map[coreunit.Name]struct{}{}
+
+	newUnitNamespace, query := s.st.InitialWatchStatementUnitInsertDeleteOnNetNode(desiredNetNodeUUID)
+	return s.watcherFactory.NewNamespaceMapperWatcher(
+		func(ctx context.Context, txn database.TxnRunner) ([]string, error) {
+			initialResults, err := query(ctx, txn)
+			if err != nil {
+				return nil, errors.Capture(err)
+			}
+			for _, result := range initialResults {
+				unitNamesOnMachineCache[coreunit.Name(result)] = struct{}{}
+			}
+			return initialResults, nil
+		},
+		func(ctx context.Context, changes []changestream.ChangeEvent) ([]changestream.ChangeEvent, error) {
+			// If there are no changes, return no changes.
+			if len(changes) == 0 {
+				return nil, nil
+			}
+
+			filteredChanges := make([]changestream.ChangeEvent, 0, len(changes))
+			for _, change := range changes {
+				unitName, err := coreunit.NewName(change.Changed())
+				if err != nil {
+					return nil, err
+				}
+				netNodeUUID, err := s.st.GetNetNodeUUIDByUnitName(ctx, unitName)
```

Comment:
Won't we need to check for the machine's life at this point? We do check the machine life in the legacy state watcher.
Maybe we should just add a TODO?

Reply:
But the behavior changes from the legacy watcher then https://github.com/juju/juju/blob/main/state/watcher.go#L1677-L1679.

Reply:
That is change in life for units, however all we are after really are INSERT and DELETE events in the table